### PR TITLE
Alistair wip

### PR DESCRIPTION
**Pull request by alistairknight**

### CODE REVIEW

File: app/routes/documents.js [hunk 1]
```diff
@@ -24,24 +24,127 @@ module.exports = [{
...
+
+          await vectorStore.addDocuments([doc])
+
+          if (/*useModel === 'ollama' &&*/ count % batchSize === 0) {
+            await vectorStore.save(`${__dirname}/../data/${saveDir}/HNSWLib_${modelName}/`)
+            console.log(`Saving ${count} / ${docs.length}`)
+          }
+        }
+
+        //if (useModel === 'ollama') {
```
Comment (BarryGeeDefra) [inline app/routes/documents.js]: can we remove these comments

File: app/data/cosine-1000-200/HNSWLib_oai/args.json [hunk 2]
```diff
@@ -0,0 +1 @@
+{"space":"cosine","numDimensions":1536}
```
Comment (BarryGeeDefra) [inline app/data/cosine-1000-200/HNSWLib_oai/args.json:1]: missing line break

File: app/data/cosine-300-70/HNSWLib_gemma/args.json [hunk 3]
```diff
@@ -0,0 +1 @@
+{"space":"cosine","numDimensions":3072}
```
Comment (BarryGeeDefra) [inline app/data/cosine-300-70/HNSWLib_gemma/args.json:1]: missing line break

File: app/data/cosine-300-70/HNSWLib_llama3/args.json [hunk 4]
```diff
@@ -0,0 +1 @@
+{"space":"cosine","numDimensions":4096}
```
Comment (BarryGeeDefra) [inline app/data/cosine-300-70/HNSWLib_llama3/args.json:1]: missing line break

File: app/data/cosine-300-70/HNSWLib_mistral/args.json [hunk 5]
```diff
@@ -0,0 +1 @@
+{"space":"cosine","numDimensions":4096}
```
Comment (BarryGeeDefra) [inline app/data/cosine-300-70/HNSWLib_mistral/args.json:1]: missing line break

File: app/data/cosine-300-70/HNSWLib_oai/args.json [hunk 6]
```diff
@@ -0,0 +1 @@
+{"space":"cosine","numDimensions":1536}
```
Comment (BarryGeeDefra) [inline app/data/cosine-300-70/HNSWLib_oai/args.json:1]: missing line break

File: app/data/cosine-500-100/HNSWLib_gemma/args.json [hunk 7]
```diff
@@ -0,0 +1 @@
+{"space":"cosine","numDimensions":3072}
```
Comment (BarryGeeDefra) [inline app/data/cosine-500-100/HNSWLib_gemma/args.json:1]: missing line break

File: app/data/cosine-500-100/HNSWLib_llama3/args.json [hunk 8]
```diff
@@ -0,0 +1 @@
+{"space":"cosine","numDimensions":4096}
```
Comment (BarryGeeDefra) [inline app/data/cosine-500-100/HNSWLib_llama3/args.json:1]: missing line break

File: app/data/cosine-500-100/HNSWLib_mistral/args.json [hunk 9]
```diff
@@ -0,0 +1 @@
+{"space":"cosine","numDimensions":4096}
```
Comment (BarryGeeDefra) [inline app/data/cosine-500-100/HNSWLib_mistral/args.json:1]: missing line break

File: app/data/cosine-500-100/HNSWLib_oai/args.json [hunk 10]
```diff
@@ -0,0 +1 @@
+{"space":"cosine","numDimensions":1536}
```
Comment (BarryGeeDefra) [inline app/data/cosine-500-100/HNSWLib_oai/args.json:1]: missing line break

File: app/data/l2-1000-200/HNSWLib_gemma/args.json [hunk 11]
```diff
@@ -0,0 +1 @@
+{"space":"l2","numDimensions":3072}
```
Comment (BarryGeeDefra) [inline app/data/l2-1000-200/HNSWLib_gemma/args.json:1]: missing line break

File: app/data/l2-1000-200/HNSWLib_llama3/args.json [hunk 12]
```diff
@@ -0,0 +1 @@
+{"space":"l2","numDimensions":4096}
```
Comment (BarryGeeDefra) [inline app/data/l2-1000-200/HNSWLib_llama3/args.json:1]: missing line break

File: app/data/l2-1000-200/HNSWLib_mistral/args.json [hunk 13]
```diff
@@ -0,0 +1 @@
+{"space":"l2","numDimensions":4096}
```
Comment (BarryGeeDefra) [inline app/data/l2-1000-200/HNSWLib_mistral/args.json:1]: missing line break

File: app/data/l2-1000-200/HNSWLib_oai/args.json [hunk 14]
```diff
@@ -0,0 +1 @@
+{"space":"l2","numDimensions":1536}
```
Comment (BarryGeeDefra) [inline app/data/l2-1000-200/HNSWLib_oai/args.json:1]: missing line break

File: app/data/l2-300-70/HNSWLib_gemma/args.json [hunk 15]
```diff
@@ -0,0 +1 @@
+{"space":"l2","numDimensions":3072}
```
Comment (BarryGeeDefra) [inline app/data/l2-300-70/HNSWLib_gemma/args.json:1]: missing line break

File: app/data/l2-300-70/HNSWLib_llama3/args.json [hunk 16]
```diff
@@ -0,0 +1 @@
+{"space":"l2","numDimensions":4096}
```
Comment (BarryGeeDefra) [inline app/data/l2-300-70/HNSWLib_llama3/args.json:1]: missing line break

File: app/data/l2-300-70/HNSWLib_mistral/args.json [hunk 17]
```diff
@@ -0,0 +1 @@
+{"space":"l2","numDimensions":4096}
```
Comment (BarryGeeDefra) [inline app/data/l2-300-70/HNSWLib_mistral/args.json:1]: missing line break

File: app/data/l2-300-70/HNSWLib_oai/args.json [hunk 18]
```diff
@@ -0,0 +1 @@
+{"space":"l2","numDimensions":1536}
```
Comment (BarryGeeDefra) [inline app/data/l2-300-70/HNSWLib_oai/args.json:1]: missing line break

File: app/data/l2-500-100/HNSWLib_gemma/args.json [hunk 19]
```diff
@@ -0,0 +1 @@
+{"space":"l2","numDimensions":3072}
```
Comment (BarryGeeDefra) [inline app/data/l2-500-100/HNSWLib_gemma/args.json:1]: missing line break

File: app/data/l2-500-100/HNSWLib_llama3/args.json [hunk 20]
```diff
@@ -0,0 +1 @@
+{"space":"l2","numDimensions":4096}
```
Comment (BarryGeeDefra) [inline app/data/l2-500-100/HNSWLib_llama3/args.json:1]: missing line break

File: app/data/l2-500-100/HNSWLib_mistral/args.json [hunk 21]
```diff
@@ -0,0 +1 @@
+{"space":"l2","numDimensions":4096}
```
Comment (BarryGeeDefra) [inline app/data/l2-500-100/HNSWLib_mistral/args.json:1]: missing line break

File: app/data/l2-500-100/HNSWLib_oai/args.json [hunk 22]
```diff
@@ -0,0 +1 @@
+{"space":"l2","numDimensions":1536}
```
Comment (BarryGeeDefra) [inline app/data/l2-500-100/HNSWLib_oai/args.json:1]: missing line break

File: app/scripts/ingest-db.js [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+      metadataColumns
+    )
+
+    let count = 0
+    for await (const doc of docs) {
+      count += 1
+
+      await vectorStore.addDocuments([doc])
+
+      if (/*useModel === 'ollama' &&*/ count % batchSize === 0) {
```
Comment (BarryGeeDefra) [inline app/scripts/ingest-db.js]: can the comments be removed?

File: app/scripts/ingest-db.js [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+        await vectorStore.save(`${__dirname}/../data/${saveDir}/HNSWLib_${modelName}/`)
+        console.log(`Saving ${count} / ${docs.length}`)
+      }
+    }
+
+    //if (useModel === 'ollama') {
```
Comment (BarryGeeDefra) [inline app/scripts/ingest-db.js]: can the comments be removed?

File: app/scripts/sql-agent.js [hunk 25]
```diff
@@ -1,17 +1,18 @@
...
+const { DataSource } = require('typeorm');
+const { model } = require('../llm/ai');
+const dbConfig = require('../config/db');
 
 
 // Ref: https://github.com/langchain-ai/langchain/blob/master/templates/sql-ollama/sql_ollama/chain.py
+//      https://js.langchain.com/v0.1/docs/use_cases/sql/quickstart/
 
 //process.env.LANGCHAIN_TRACING_V2 = true
 //process.env.LANGCHAIN_API_KEY = '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'
```
Comment (BarryGeeDefra) [inline app/scripts/sql-agent.js:18]: can the comments be removed?

File: app/scripts/vector-test.js [hunk 26]
```diff
@@ -0,0 +1,27 @@
...
+(async () => {
+  const modelName = useModel === 'ollama' ? ollamaModel : 'oai'
+
+  console.log('modelName:', modelName)
+  console.log('saveDir:', saveDir)
+
+  let testStr = 'who established the Inshore Fisheries and Conservation Authorities'
+  const vectorStore = await getVectorStore('load', useModel)
+
+  /*
```
Comment (BarryGeeDefra) [inline app/scripts/vector-test.js]: can the comments be removed?

File: app/scripts/webscrape-db.js [hunk 27]
```diff
@@ -101,12 +101,13 @@ const importData = async (filename) => {
 
 (async () => {
   const url = 'https://sciencesearch.defra.gov.uk/'
-  const filename = 'pages1.json'
+  const directoryPath = `${__dirname}/../data/`
+  const fileName = 'sciencesearch.defra.gov.uk_evaluations.json'
 
-  //const fields = await inspect(filename)
```
Comment (BarryGeeDefra) [inline app/scripts/webscrape-db.js:106]: can the comments be removed?

File: app/views/query.njk [hunk 28]
```diff
@@ -12,21 +13,98 @@
...
+            {% if item.pageContent %}
+              {{ item.pageContent }}{% if item.pageNumber %}, page {{ item.pageNumber }}{% endif %}
+            {% endif %}
+          </li>
+          {% endfor %}
+        </ol>
+      {% endif %}
     </div>
   </div>
 {% endblock %}
```
Comment (BarryGeeDefra) [inline app/views/query.njk]: missing line break

**Review by BarryGeeDefra — COMMENTED** [review]
Just a few minor changes, comment removals and missing line breaks that would prevent a merge. Other than that all good.

File: app/data/cosine-1000-200/HNSWLib_oai/args.json (hunk 2 above)
Comment (alistairknight) [reply app/data/cosine-1000-200/HNSWLib_oai/args.json:1]: This file is auto-generated

File: app/data/cosine-300-70/HNSWLib_llama3/args.json (hunk 4 above)
Comment (alistairknight) [reply app/data/cosine-300-70/HNSWLib_llama3/args.json:1]: This file is auto-generated

File: app/data/cosine-300-70/HNSWLib_mistral/args.json (hunk 5 above)
Comment (alistairknight) [reply app/data/cosine-300-70/HNSWLib_mistral/args.json:1]: This file is auto-generated

File: app/data/cosine-500-100/HNSWLib_gemma/args.json (hunk 7 above)
Comment (alistairknight) [reply app/data/cosine-500-100/HNSWLib_gemma/args.json:1]: This file is auto-generated

File: app/data/cosine-500-100/HNSWLib_mistral/args.json (hunk 9 above)
Comment (alistairknight) [reply app/data/cosine-500-100/HNSWLib_mistral/args.json:1]: This file is auto-generated

File: app/data/l2-1000-200/HNSWLib_gemma/args.json (hunk 11 above)
Comment (alistairknight) [reply app/data/l2-1000-200/HNSWLib_gemma/args.json:1]: This file is auto-generated

File: app/data/l2-1000-200/HNSWLib_llama3/args.json (hunk 12 above)
Comment (alistairknight) [reply app/data/l2-1000-200/HNSWLib_llama3/args.json:1]: This file is auto-generated

File: app/data/l2-1000-200/HNSWLib_mistral/args.json (hunk 13 above)
Comment (alistairknight) [reply app/data/l2-1000-200/HNSWLib_mistral/args.json:1]: This file is auto-generated

File: app/data/l2-1000-200/HNSWLib_oai/args.json (hunk 14 above)
Comment (alistairknight) [reply app/data/l2-1000-200/HNSWLib_oai/args.json:1]: This file is auto-generated

File: app/data/l2-300-70/HNSWLib_gemma/args.json (hunk 15 above)
Comment (alistairknight) [reply app/data/l2-300-70/HNSWLib_gemma/args.json:1]: This file is auto-generated

File: app/data/l2-300-70/HNSWLib_llama3/args.json (hunk 16 above)
Comment (alistairknight) [reply app/data/l2-300-70/HNSWLib_llama3/args.json:1]: This file is auto-generated

File: app/data/l2-300-70/HNSWLib_mistral/args.json (hunk 17 above)
Comment (alistairknight) [reply app/data/l2-300-70/HNSWLib_mistral/args.json:1]: This file is auto-generated

File: app/data/l2-300-70/HNSWLib_oai/args.json (hunk 18 above)
Comment (alistairknight) [reply app/data/l2-300-70/HNSWLib_oai/args.json:1]: This file is auto-generated

File: app/data/l2-500-100/HNSWLib_gemma/args.json (hunk 19 above)
Comment (alistairknight) [reply app/data/l2-500-100/HNSWLib_gemma/args.json:1]: This file is auto-generated

File: app/data/l2-500-100/HNSWLib_llama3/args.json (hunk 20 above)
Comment (alistairknight) [reply app/data/l2-500-100/HNSWLib_llama3/args.json:1]: This file is auto-generated

File: app/data/l2-500-100/HNSWLib_mistral/args.json (hunk 21 above)
Comment (alistairknight) [reply app/data/l2-500-100/HNSWLib_mistral/args.json:1]: This file is auto-generated

File: app/data/l2-500-100/HNSWLib_oai/args.json (hunk 22 above)
Comment (alistairknight) [reply app/data/l2-500-100/HNSWLib_oai/args.json:1]: This file is auto-generated

File: app/scripts/sql-agent.js (hunk 25 above)
Comment (alistairknight) [reply app/scripts/sql-agent.js:18]: I'd prefer to leave references in

File: app/data/cosine-500-100/HNSWLib_oai/args.json (hunk 10 above)
Comment (alistairknight) [reply app/data/cosine-500-100/HNSWLib_oai/args.json:1]: This file is auto-generated

File: app/data/cosine-500-100/HNSWLib_llama3/args.json (hunk 8 above)
Comment (alistairknight) [reply app/data/cosine-500-100/HNSWLib_llama3/args.json:1]: This file is auto-generated

File: app/data/cosine-300-70/HNSWLib_oai/args.json (hunk 6 above)
Comment (alistairknight) [reply app/data/cosine-300-70/HNSWLib_oai/args.json:1]: This file is auto-generated

File: app/data/cosine-300-70/HNSWLib_gemma/args.json (hunk 3 above)
Comment (alistairknight) [reply app/data/cosine-300-70/HNSWLib_gemma/args.json:1]: This file is auto-generated

File: app/routes/documents.js (hunk 1 above)
Comment (alistairknight) [reply app/routes/documents.js]: Sorted

File: app/scripts/ingest-db.js (hunk 24 above)
Comment (alistairknight) [reply app/scripts/ingest-db.js]: Sorted

File: app/scripts/vector-test.js (hunk 26 above)
Comment (alistairknight) [reply app/scripts/vector-test.js]: Sorted

File: app/scripts/webscrape-db.js (hunk 27 above)
Comment (alistairknight) [reply app/scripts/webscrape-db.js:106]: Sorted

File: app/scripts/ingest-db.js (hunk 23 above)
Comment (alistairknight) [reply app/scripts/ingest-db.js]: Sorted

File: app/views/query.njk (hunk 28 above)
Comment (alistairknight) [reply app/views/query.njk]: Sorted

**Review by BarryGeeDefra — APPROVED** [review]
LGTM

**Pull request merged**